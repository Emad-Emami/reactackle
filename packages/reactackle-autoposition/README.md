# Reactackle-autoposition


**Part of [Reactackle](https://www.npmjs.com/package/reactackle) - a new open-source components library based on [React](https://facebook.github.io/react/) and [Styled Components](https://www.styled-components.com).**

## Installation

**Using NPM:**
```
npm install reactackle-autoposition --save
```

**Using Yarn:**
```
yarn add reactackle-autoposition
```
[Component Demo](http://reactackle-docs.braincrumbs.io/#/autoposition/demo)
