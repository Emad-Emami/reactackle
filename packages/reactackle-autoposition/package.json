{
  "name": "reactackle-autoposition",
  "version": "0.0.3-beta.1",
  "description": "Autoposition component of the Reactackle ui library",
  "scripts": {
    "prepublishOnly": "rimraf ./dist && npx babel ./src --out-dir ./dist"
  },
  "author": "braincrumbs",
  "license": "MIT",
  "keywords": [
    "react",
    "components",
    "reactackle"
  ],
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "dependencies": {
    "lodash.pick": "^4.4.0",
    "lodash.throttle": "^4.1.1",
    "prop-types": "^15.5.10",
    "react-portal": "^3.1.0"
  },
  "peerDependencies": {
    "react": "^15.6.1",
    "reactackle-core": "^0.0.3-beta.1"
  },
  "devDependencies": {
    "babel-cli": "^6.24.1",
    "react-test-renderer": "^15.6.1",
    "reactackle-core": "^0.0.3-beta.1",
    "rimraf": "^2.6.1"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/bcrumbs/reactackle.git"
  },
  "bugs": "https://github.com/bcrumbs/reactackle/issues",
  "homepage": "https://reactackle.braincrumbs.io/"
}
