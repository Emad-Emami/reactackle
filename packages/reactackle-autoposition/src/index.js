import AutoPosition, { recalculateAllAutoPosition } from './AutoPosition';

export { AutoPosition, recalculateAllAutoPosition };
