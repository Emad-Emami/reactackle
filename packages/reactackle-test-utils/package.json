{
  "name": "reactackle-test-utils",
  "version": "1.0.0-beta.7",
  "description": "Test utils of the Reactackle ui library",
  "scripts": {
    "prepublishOnly": "rimraf ./dist && npx babel ./src --out-dir ./dist"
  },
  "author": "braincrumbs",
  "license": "MIT",
  "keywords": [
    "react",
    "components",
    "reactackle"
  ],
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "dependencies": {
    "enzyme": "^3.1.0",
    "enzyme-to-json": "^3.1.4",
    "react-test-renderer": "^16.0.0"
  },
  "peerDependencies": {
    "react": "^16.0.0"
  },
  "devDependencies": {
    "babel-cli": "^6.24.1",
    "rimraf": "^2.6.1"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/bcrumbs/reactackle.git"
  },
  "bugs": "https://github.com/bcrumbs/reactackle/issues",
  "homepage": "https://reactackle.braincrumbs.io/"
}
