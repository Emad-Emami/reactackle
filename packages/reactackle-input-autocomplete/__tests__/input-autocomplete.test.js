'use strict';

import React from 'react';
import renderer from 'react-test-renderer';
import { AutoPosition } from 'reactackle-autoposition';
import {
  composeReactackleTheme,
  keyboardCodes,
} from 'reactackle-core';
import { findByType, findPortal } from 'reactackle-test-utils';

import InputAutocompleteWithHOC from '../src/InputAutocomplete';

const InputAutocomplete = InputAutocompleteWithHOC.withoutHOC;
jest.mock('react-dom');

const options = [
  { value: 'Option 1' },
  { value: 'Option 1.1' },
  { value: 'Option 2' },
  { value: 'Option 2.1' },
];

const map = {};

window.addEventListener = jest.fn((event, cb) => {
  map[event] = cb;
});

describe('<InputAutocomplete/>', () => {
  it('renders correctly with default props', () => {
    const tree = renderer.create(<InputAutocomplete />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly if set options', () => {
    const tree = renderer
      .create(<InputAutocomplete options={options} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly if set options and init value', () => {
    const tree = renderer
      .create(<InputAutocomplete options={options} defaultValue="Opt" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly if set options, init value and stateless', () => {
    const tree = renderer
      .create(<InputAutocomplete options={options} value="Opt" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly if set options and call change option amount', () => {
    const component = renderer.create(
      <InputAutocomplete options={options} value="Opt" />,
    );

    component.getInstance().componentWillReceiveProps({ options: [] });
    expect(component.toJSON()).toMatchSnapshot();

    component.getInstance().componentWillReceiveProps({ options });
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders correctly if set options and call onChange input', () => {
    const component = renderer.create(<InputAutocomplete options={options} />),
      value = 'Opt';

    const input = findByType(component.toJSON(), 'input');
    input.props.onChange({ target: { value } });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders correctly if set options and complete first option', () => {
    const theme = composeReactackleTheme();
    const component = renderer.create(
        <InputAutocomplete theme={theme} options={options} />,
      ),
      value = 'Opt';
    const setScrollMock = jest.fn();

    component.getInstance()._setScrollPositionByOptionKey = setScrollMock;

    const input = findByType(component.toJSON(), 'input');
    input.props.onChange({ target: { value } });
    input.props.onFocus();

    expect(component.toJSON()).toMatchSnapshot();

    map.keydown({
      keyCode: keyboardCodes.UP,
      preventDefault: jest.fn(),
    });
    expect(component.toJSON()).toMatchSnapshot();
    expect(setScrollMock.mock.calls.length).toEqual(1);

    map.keydown({
      keyCode: keyboardCodes.DOWN,
      preventDefault: jest.fn(),
    });

    expect(component.toJSON()).toMatchSnapshot();
    expect(setScrollMock.mock.calls.length).toEqual(2);

    component.getInstance().setState({ highlightedOption: 2 });
    map.keydown({
      keyCode: keyboardCodes.ENTER,
      preventDefault: jest.fn(),
    });

    expect(component.toJSON()).toMatchSnapshot();
    input.props.onChange({ target: { value } });
    input.props.onFocus();

    map.keydown({
      keyCode: keyboardCodes.ENTER,
      preventDefault: jest.fn(),
    });

    expect(component.toJSON()).toMatchSnapshot();

    input.props.onFocus();
    map.keydown({
      keyCode: keyboardCodes.ESC,
      preventDefault: jest.fn(),
    });

    expect(component.toJSON()).toMatchSnapshot();

    map.keydown({
      keyCode: keyboardCodes.ESC,
      preventDefault: jest.fn(),
    });
  });

  it('handles unmount', () => {
    const component = renderer.create(
      <InputAutocomplete stateless defaultValue="" options={options} />,
    );
    expect(() => component.unmount()).not.toThrow();
  });

  it('handles blur', () => {
    const mockBlur = jest.fn();
    const component = renderer.create(
      <InputAutocomplete onBlur={mockBlur} options={options} />,
    );
    const input = findByType(component.toJSON(), 'input');
    input.props.onBlur();
    expect(mockBlur).toBeCalled();
  });

  it('handles focus', () => {
    const mockFocus = jest.fn();
    const component = renderer.create(
      <InputAutocomplete onFocus={mockFocus} options={options} />,
    );
    const input = findByType(component.toJSON(), 'input');
    component.getInstance().setState({ showOptions: false });
    input.props.onFocus();
    expect(mockFocus).toBeCalled();
  });

  it('handles option mouse events', () => {
    const component = renderer.create(
      <InputAutocomplete
        stateless
        options={[
          {
            value: 'Foo',
          },
        ]}
      />,
    );
    const input = findByType(component.toJSON(), 'input');
    input.props.onFocus();

    const portal = findPortal(component, AutoPosition);
    const option = findByType(portal.toJSON(), 'li');
    const mockEvent = {
      nativeEvent: {
        target: {
          getAttribute: () => 0,
        },
        relatedTarget: {
          getAttribute: () => 0,
        },
      },
    };
    option.props.onMouseEnter(mockEvent);
    option.props.onMouseLeave(mockEvent);
    option.props.onClick(mockEvent);
  });
});
