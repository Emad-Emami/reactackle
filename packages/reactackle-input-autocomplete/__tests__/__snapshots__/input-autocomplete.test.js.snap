// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputAutocomplete/> renders correctly if set options 1`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA hHuqwi"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-bwzfXH iprzAE"
              disabled={false}
              id="TextField_1"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and call change option amount 1`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA hHuqwi"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-ifAKCX iibpDV"
              disabled={false}
              id="TextField_4"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Opt"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and call change option amount 2`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA hHuqwi"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-ifAKCX iibpDV"
              disabled={false}
              id="TextField_4"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Opt"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and call onChange input 1`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA hHuqwi"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-EHOje doiYmN"
              disabled={false}
              id="TextField_5"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Opt"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and complete first option 1`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA dfUXvC"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-bZQynM dUGAxg"
              disabled={false}
              id="TextField_6"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Opt"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and complete first option 2`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA dfUXvC"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-bZQynM dUGAxg"
              disabled={false}
              id="TextField_6"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Option 2.1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and complete first option 3`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA dfUXvC"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-bZQynM dUGAxg"
              disabled={false}
              id="TextField_6"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Option 1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and complete first option 4`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA dfUXvC"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-bZQynM dUGAxg"
              disabled={false}
              id="TextField_6"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Option 2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and complete first option 5`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA dfUXvC"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-bZQynM dUGAxg"
              disabled={false}
              id="TextField_6"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Opt"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and complete first option 6`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA dfUXvC"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-bZQynM dUGAxg"
              disabled={false}
              id="TextField_6"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Opt"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options and init value 1`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA hHuqwi"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-htpNat brhVmQ"
              disabled={false}
              id="TextField_2"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Opt"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly if set options, init value and stateless 1`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA hHuqwi"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-bxivhb fcorOR"
              disabled={false}
              id="TextField_3"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value="Opt"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InputAutocomplete/> renders correctly with default props 1`] = `
<div
  className="InputAutocompleteStyled__InputAutocompleteStyled-emtqUB joCwNK"
>
  <div>
    <div
      className="TextFieldStyled__TextFieldStyled-LQiMh bRFqzw"
    >
      <div
        className="TextFieldContentBoxStyled__TextFieldContentBoxStyled-eUFbzZ iDAUec"
      >
        <div
          className="TextFieldRowStyled__TextFieldRowStyled-iKawDw hjaZvM"
          disabled={false}
        >
          <div
            className="TextFieldGroupStyled__TextFieldGroupStyled-cIZJDA hHuqwi"
            disabled={false}
          >
            <input
              autoComplete="off"
              className="sc-bdVaJa bteGOT"
              disabled={false}
              id="TextField_0"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              placeholder=""
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
